src/memory_manager.rs

```diff
@@ -287,13 +393,99 @@ impl<M: Memory> MemoryManagerInner<M> {
             bucket_size_in_pages: header.bucket_size_in_pages,
             memory_sizes_in_pages: header.memory_sizes_in_pages,
             memory_buckets,
+            freed_buckets: BTreeSet::new(),
+        }
+    }
+
+    fn load_layout_v2(memory: M, header: Header) -> Self {
+        const BYTE_SIZE_IN_BITS: usize = 8;
+        const BUCKETS_INDEX_SIZE_IN_BYTES: usize =
+            (MAX_NUM_BUCKETS as usize * BUCKET_ID_LEN_IN_BITS + (BYTE_SIZE_IN_BITS - 1))
+                / BYTE_SIZE_IN_BITS;
+        let mut buckets = vec![0; BUCKETS_INDEX_SIZE_IN_BYTES];
+        memory.read(bucket_indexes_offset().get(), &mut buckets);
+
+        let buckets_decompressed = bytes_to_bucket_indexes(&buckets);
+
+        let mut memory_buckets = BTreeMap::new();
+
+        let mut last_occupied_bucket: i32 = -1;
+
+        let mut j = 0;
+        for (memory, memory_size_in_pages) in header.memory_sizes_in_pages.into_iter().enumerate() {
+            let memory_size_in_buckets =
+                (memory_size_in_pages + header.bucket_size_in_pages as u64 - 1)
+                    / header.bucket_size_in_pages as u64;
+            let mut vec_buckets = vec![];
+            for _ in 0..memory_size_in_buckets {
+                let bucket = buckets_decompressed[j];
+                last_occupied_bucket = std::cmp::max(bucket.0 as i32, last_occupied_bucket);
+                vec_buckets.push(bucket);
+                j += 1;
+            }
+            memory_buckets
+                .entry(MemoryId(memory as u8))
+                .or_insert(vec_buckets);
+        }
+
+        let mem_mgr = if last_occupied_bucket == -1 {
+            Self {
+                memory,
+                allocated_buckets: header.num_allocated_buckets,
+                bucket_size_in_pages: header.bucket_size_in_pages,
+                memory_sizes_in_pages: header.memory_sizes_in_pages,
+                memory_buckets,
+                freed_buckets: BTreeSet::new(),
+            }
+        } else {
```

Comment:
I believe here is incorrect. Here's a test case that fails:

```
    #[test]
    fn freed_memories_are_tracked() {
        let mem = make_memory();
        let mut mem_mgr = MemoryManager::init(mem.clone());
        mem_mgr.get(MemoryId(0)).grow(1);
        mem_mgr.get(MemoryId(1)).grow(1);
        mem_mgr.free(MemoryId(0));

        let mem_mgr = MemoryManager::init(mem);
        assert_eq!(
            mem_mgr.inner.borrow().freed_buckets,
            maplit::btreeset! { BucketId(0) }
        );
    }
```

Reply:
Great point! I fixed it now, that was the problem with loading freed buckets when the Bucket ID is 0. PTAL.